fix blocks.py, no jinja2, nvm&npm install

dnsl48 opened the pull request:
Fix blocks.py implementation
No Jinja2 anymore
No custom filters
No custom template tags
Fix Makefile / add nvm install and npm install to "make init"

dnsl48 committed:
Fix blocks.py import paths
Add Jinja2 as a requirement because of filters.py
Fix Makefile / add nvm install and npm install to "make init"

thibaudcolas commented on Makefile:
```diff
@@ -36,6 +36,8 @@ clean-pyc: ## Remove Python file artifacts.
 
 dist: ## Compile the JS and CSS for release.
+	nvm install || echo "nvm is not available"
+	npm install
```
This should go in the init target – `dist` is ran on every `publish`.

thibaudcolas commented on setup.py:
```diff
@@ -14,6 +14,7 @@
 
 install_requires = [
     'draftjs_exporter==0.7.0',
+    'jinja2>=2.9.5,<3'  # TODO: remove when it's not necessary (https://github.com/springload/wagtaildraftail/issues/13)
```
As far as I know, at the moment this package supports both Jinja2 and Django templates. Is it a good idea to force the install of Jinja2 in here?

I also realise that jinja2 is required in `filters.py`, and somehow I think I still got this to work in the test site in `tests` without having it installed. What's going on?

loicteixeira replied:
Wagtail implements everything for `Django Templates` by default and provide extra `Jinja2` tags in a [separate class](https://github.com/wagtail/wagtail/blob/1874b8b259e1295370b1c2122966d96668db7eec/wagtail/wagtailcore/jinja2tags.py#L10), which is not imported from anywhere by default. Then, if an implementation uses `Jinja2`, it can be [configured](https://github.com/wagtail/wagtail/blob/85de048a890aec4be3bcbf3af8a9ecfbf467d34a/wagtail/tests/settings.py#L69) to be included (and then it's up to the developer to make sure the right dependencies are installed). I think we should follow the same idea and not require `Jinja2` as a dependency for this package.

dnsl48 replied:
Well, as far as we use WagtaildraftTextBlock there is no way to avoid having Jinja2 as a dependency. However, it appears we don't really need Jinja2 since we only use MarkSafe.Markup  class. Furthermore, we probably don't even need MarkSafe as a dependency, because we apparently don't use its features - I'll try to go with django.utils.safestring.mark_safe instead

loicteixeira replied:
Wagtail's `rich_text` filter [does use mark_safe](https://github.com/wagtail/wagtail/blob/master/wagtail/wagtailcore/templatetags/wagtailcore_tags.py#L53-L62) so the `draft_text` should do the same.

loicteixeira replied:
Oh, apparently the equivalent [Django Template Tag](https://github.com/springload/wagtaildraftail/blob/758254139cebb6e6c3d3d8db731615772a2f535e/wagtaildraftail/templatetags/draftail_tags.py) has already been created so `DraftailTextBlock` should use this instead. Then if you make sure that `wagtaildraftail.filters` is never imported (and later converted to something similar to [this](https://github.com/wagtail/wagtail/blob/1874b8b259e1295370b1c2122966d96668db7eec/wagtail/wagtailcore/jinja2tags.py#L10)) and the dependency can be removed.

loicteixeira replied:
@dnsl48 found that if we create a template node which inherit from [RichText](https://github.com/wagtail/wagtail/blob/master/wagtail/wagtailcore/rich_text.py#L186) and provide our own methods, the [rich_text](https://github.com/wagtail/wagtail/blob/master/wagtail/wagtailcore/templatetags/wagtailcore_tags.py#L54) filter will transparently consume Draft JSON data which means we can remove the need for template tags in this project (which also solve what was express in #13).

thibaudcolas commented on wagtaildraftail/blocks.py:
```diff
@@ -6,8 +6,8 @@
-import forms
-from filters import draft_text
+from wagtaildraftail.forms import SerializedJSONField
+from wagtaildraftail.filters import draft_text
```
Travis is complaining (well, isort) because `filters` should go before `form`.

thibaudcolas reviewed (commented):
Some things I'm not sure of, some things that need fixing for the linting to pass.

I'm really not sure what to do about Jinja2 vs Django Templates – hopefully @loicteixeira will know.

dnsl48 committed:
blocks.py / Fix import ordering

dnsl48 committed:
Makefile / nvm install && npm install to be in the init target

dnsl48 committed:
DraftText(RichText) implementation
New DraftailTextBlock using new DraftText implementation
Remove custom filters
Remove custom template tags

Fixes #13

dnsl48 commented:
please don't merge yet - I'm going to check DraftailTextField yet

dnsl48 committed:
DraftailTextField to produce DraftText

dnsl48 commented:
Ok, now it seems finished. Gentlemen, please review @loicteixeira, @thibaudcolas

loicteixeira commented on wagtaildraftail/draft_text.py:
```diff
@@ -0,0 +1,17 @@
+        self.exporter = HTML(get_exporter_config())
+
+    def __html__(self):
+        return self.exporter.render(json.loads(self.source))
```
The base constructor does `self.source = (source or '')` which means that `json.loads` will raise a `ValueError`. We should either change the constructor to `self.source = (source or '{}')` or handle the exception here.

loicteixeira commented:
Also build is failing :(

dnsl48 commented:
yeah, seems preliminary, sorry for that :/

dnsl48 committed:
Fix DraftText usage around database, form widget and live page

dnsl48 committed:
Fix DraftailTextField prep_value generation for Django>1.8

dnsl48 commented:
Ok, this time tested on both Django `1.8` and `1.10`
@loicteixeira, @thibaudcolas please, carry on the review

loicteixeira reviewed (approved):
It looks good to me 👌
